### DISCUSSION TITLE
How to manage environment variables?

### DISCUSSION THREAD

**Discussion by v-raja** [nandorojo/solito#31] [Q&A] [answered]
I'm migrating a Next.js app to a RN + Next app using the solito blank monorepo. I was wondering where I should store my environment variables so that I can use `process.env.<KEY>` in my code and it would work across both platforms?

**Comment by raulghm** [comment]
same question :m

**Comment by nandorojo** [comment]
I have a solution to this. I’ll write here when I have some time.

**Reply by rodbs** [reply]
I've found this 2 solutions, but I haven't been able to integrate them in the `solito` architecture
https://medium.com/swlh/how-to-properly-use-environment-variables-in-an-expo-react-native-app-7ab852590b30
https://github.com/egislook/react-native-expand-dotenv

@nandorojo  It'd be highly appreciated if you could explain the architecture of the monorepo: how packages are read and how NextJs is 'integrated' into `expo web` , into `expo` ....

**Reply by nandorojo** [reply]
See my answer here: https://github.com/nandorojo/solito/discussions/31#discussioncomment-2576208

**Comment by nandorojo** [comment, marked as answer]
While I don't have a ton of time to write a full guide or example currently, I realized I explained this over a discord chat before already. So I'm going to drop screenshots here of that convo. Hopefully this is enough to explain it all.

**Please consider these instructions as-is. You should never commit any secrets to git or put them in any env files using this approach. You can use secrets on Vercel and on EAS. I'm sharing this as a courtesy, but you are in charge of your own security.**

<img width="1047" alt="Screen Shot 2022-04-15 at 2 51 07 PM" src="https://user-images.githubusercontent.com/13172299/163614619-c800c93c-f9c4-4970-89ec-4534c64dabbd.png">

<img width="1035" alt="Screen Shot 2022-04-15 at 2 51 21 PM" src="https://user-images.githubusercontent.com/13172299/163614613-e9d85062-3962-4b2c-9d62-44210d648bf4.png">

<img width="1034" alt="Screen Shot 2022-04-15 at 2 51 34 PM" src="https://user-images.githubusercontent.com/13172299/163614608-8cf6fdae-3a34-4ffb-8083-52d4684ccd3a.png">

<img width="1036" alt="Screen Shot 2022-04-15 at 2 51 44 PM" src="https://user-images.githubusercontent.com/13172299/163614603-c552d9ef-995f-49a2-8a34-a509e4227d22.png">

![image](https://user-images.githubusercontent.com/13172299/163616092-9da03f57-0f55-4bad-bd62-917fd892c649.png)

<img width="964" alt="Screen Shot 2022-04-15 at 2 51 53 PM" src="https://user-images.githubusercontent.com/13172299/163614595-6b7a4108-2265-4e28-8e68-2bdbc08fc4f3.png">

![image](https://user-images.githubusercontent.com/13172299/163615838-703b8493-981c-4f00-8370-1af451a78b9d.png)

## Small Update

As of newer expo SDKs, I had to add this to my `eas.json` build profiles to make the above approach work, which removes Expo's built-in env variable support.

```json
"env": {
  "EXPO_NO_DOTENV": "1",
  "EXPO_NO_CLIENT_ENV_VARS": "1"
}
```

**Reply by nandorojo** [reply]
Some of the things are slightly outdated, such as the plugins in the babel config.

Also, rather than `api.cache(false)`, you could use `app.cache.invalidate` in the `babel.config.js`:

```js
module.exports = function (api) {
  const APP_ENV = process.env.APP_ENV || 'release'
  api.cache.invalidate(() => APP_ENV)
}
```

**Reply by rodbs** [reply]
Ok, thanks!! A lot of info! :)

**Reply by rodbs** [reply]
@nandorojo 
I've tried your scripts and I've managed to make it work fro NextJS, but not for expo. I've tried different configurations and always get these two errors.

```
Error: NEXT_PUBLIC_SUPABASE_URL and NEXT_PUBLIC_SUPABASE_ANON_KEY env variables are required!  
at C:\sources\igls-solito\node_modules\react-native\Libraries\LogBox\LogBox.js:149:8 in registerError
at C:\sources\igls-solito\node_modules\react-native\Libraries\LogBox\LogBox.js:60:8 in errorImpl
at C:\sources\igls-solito\node_modules\react-native\Libraries\LogBox\LogBox.js:34:4 in console.error
at C:\sources\igls-solito\node_modules\expo\build\environment\react-native-logs.fx.js:27:4 in error
at C:\sources\igls-solito\node_modules\react-native\Libraries\Core\ExceptionsManager.js:104:6 in reportException
at C:\sources\igls-solito\node_modules\react-native\Libraries\Core\ExceptionsManager.js:172:19 
in handleException
at C:\sources\igls-solito\node_modules\react-native\Libraries\Core\setUpErrorHandling.js:24:6 in handleError
at C:\sources\igls-solito\node_modules\@react-native\polyfills\error-guard.js:49:36 in ErrorUtils.reportFatalError
at C:\sources\igls-solito\node_modules\metro-runtime\src\polyfills\require.js:204:6 in guardedLoadModule
at http://192.168.0.101:19000/index.bundle?platform=android&dev=true&hot=false&strict=false&minify=false:322998:3 in global code

Invariant Violation: "main" has not been registered. This can happen if:
* Metro (the local dev server) is run from the wrong folder. Check if Metro is running, stop it and restart it in the current project.
* A module failed to load due to an error and `AppRegistry.registerComponent` wasn't called.   
at C:\sources\igls-solito\node_modules\react-native\Libraries\LogBox\LogBox.js:149:8 in registerError
at C:\sources\igls-solito\node_modules\react-native\Libraries\LogBox\LogBox.js:60:8 in errorImpl
at C:\sources\igls-solito\node_modules\react-native\Libraries\LogBox\LogBox.js:34:4 in console.error
at C:\sources\igls-solito\node_modules\expo\build\environment\react-native-logs.fx.js:27:4 in error
at C:\sources\igls-solito\node_modules\react-native\Libraries\Core\ExceptionsManager.js:104:6 in reportException
at C:\sources\igls-solito\node_modules\react-native\Libraries\Core\ExceptionsManager.js:172:19 
in handleException
at C:\sources\igls-solito\node_modules\react-native\Libraries\Core\setUpErrorHandling.js:24:6 in handleError
at C:\sources\igls-solito\node_modules\@react-native\polyfills\error-guard.js:49:36 in ErrorUtils.reportFatalError
at C:\sources\igls-solito\node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:367:8 in __guard
at C:\sources\igls-solito\node_modules\react-native\Libraries\BatchedBridge\MessageQueue.js:112:4 in callFunctionReturnFlushedQueue
```
`package.json`
``` 
 "scripts": {
    "native": "cd apps/expo && cross-env APP_ENV=development expo start",
    "web": "cd apps/next && cross-env  APP_ENV=development yarn next"
```
This is my babel conf
```
const path = require('path')

module.exports = function (api) {
  const APP_ENV = process.env.APP_ENV || 'development'
  api.cache.invalidate(() => APP_ENV)

  // const envPath = path.resolve(
  //   __dirname,
  //   `../../.vercel`,
  //   `.env.${APP_ENV}.local`
  // )

  const envPath = path.resolve(
    __dirname,
    'apps',
    'expo',
    `.env.development.local`
  )

  require('dotenv').config({ path: envPath })

  const isDev = APP_ENV.startsWith('dev')

  return {
    presets: [
      '@expo/next-adapter/babel',
      ['babel-preset-expo', { jsxRuntime: 'automatic' }],
    ],
    plugins: [['inline-dotenv', { path: envPath }]],
  }
}
``` 
(I've tried different locations for the env file)

How can I debug this babel conf? I cannot see even if it's taking the path right ...
What is the `filter(Boolean)` at the end of your example for?

By the way, I've also generated the `eas.json`, but is it necessary for local dev?
```
  "build": {
    "development": {
      "developmentClient": true,
      "distribution": "internal",
      "env": {
        "APP_ENV": "development"
      }
    },
```
Thanks!

**Reply by nandorojo** [reply]
your babel config looks wrong. notice that dirname is already apps/expo.

**Reply by nandorojo** [reply]
eas.json is not used for local dev.

you should do the path the way my screenshot did, assuming your env files are at the root.

**Reply by rodbs** [reply]
I've checked different combinations of paths and it fails.
But I think it's because the Supabase provider is trying to access the env variables before they have been loaded. Might it be possible?  Is babel  is supposed to be executed  before rendering? Or should I implement some kind of loading status (setLoading...) to wait for the env variables to be loaded? how? thx!

**Reply by nandorojo** [reply]
but why aren’t you doing the combination I used?

**Reply by nandorojo** [reply]
babel runs before the app in order to bundle the code

**Reply by nandorojo** [reply]
your path looks completely wrong. it should point to the location of the `.env` files, relative to the location of `apps/expo`

**Reply by rodbs** [reply]
I used also your combination, and it failed (it's the first one I used)
I think it'ss failing because of this in `provider/index.js`
```
import { UserProvider } from '@supabase/supabase-auth-helpers/react'
import { supabaseClient } from '@supabase/supabase-auth-helpers/nextjs'

```
It tries to load the env variables in the background, it doesn't find them and it fails.

**Reply by nandorojo** [reply]
what does the folder structure look like? this is the babel config inside of `apps/expo` right?

could you try `expo start —clear`?

**Reply by rodbs** [reply]
Finally I think I make it work. I think `expo start —clear` fix it! What a nightmare! :)
Thanks a lot!

One last question.
In `next.config.ts` you inject/filter just the public variables (NEXT_PUBLIC_) (because the are the ones to inject in the client?)
But what about the private variables? how are they available to the app/vercel? babel takes charges of it?

**Reply by nandorojo** [reply]
for the secrets, you can use `.env.local` in `apps/next` locally (this must be in gitignore!)

and then on vercel you add them to the env variables in your project settings.

**Reply by RicardoPBarbosa** [reply]
Hey, I've configured following your instructions, it's working for Next.js using `process.env.*`, but how do I make it work on Expo? It just doesn't show anything.
I got a shared folder of firebase functions to be accessed by both apps and I need to access these environment variables there to initialize firebase.

**Reply by nandorojo** [reply]
## Update

As of newer expo SDKs, I had to add this to my `eas.json` build profiles to make the above approach work, which removes Expo's built-in env variable support.

```json
"env": {
  "EXPO_NO_DOTENV": "1",
  "EXPO_NO_CLIENT_ENV_VARS": "1"
}
```

**Reply by TechchazeInc** [reply]
> Hey, I've configured following your instructions, it's working for Next.js using `process.env.*`, but how do I make it work on Expo? It just doesn't show anything. I got a shared folder of firebase functions to be accessed by both apps and I need to access these environment variables there to initialize firebase.

Hey did you figured this out, i am also facing same issue.....

**Comment by anmoltiwari00** [comment]
hey @nandorojo possible to list the extra steps required for native? would be really helpful. We have a setup with RNCli and Next. Thanks!

**Reply by nandorojo** [reply]
The babel steps should all work with native

**Reply by anmoltiwari00** [reply]
yup the babel steps make sense...but you have mentioned in step no. 5 that with native there are some extra steps to run it

**Reply by nandorojo** [reply]
the extra steps are the ones mentioned before, where I create a different dev client with each env file. this uses EAS profiles. EAS works with react native cli too, so you could use that there!

**Comment by sarat1669** [comment]
Can this be added to the solito monorepo by default?

**Reply by artiphishle** [reply]
the dotenv approach described above seems a nice way to load the ENV's, however it's the first thing you expect to already be available within the starter. Please add this great code to the starter, thanks.

(not that I'm too lazy, but now everyone first has to investigate, find this page, then add it manually... feels like a showstopper right at the beginning)

**Reply by nandorojo** [reply]
First step may be to document it. One concern with including it in the root is that people will expect it to work like next.js, where you can include secrets.
